### ISSUE TITLE
No more than ~500 GAs with GET

### ISSUE THREAD

**Issue by Makki1** [CometVisu/CometVisu#661]
Opening this issue just to track it: HTTP GET is size-limited, so some users run into a limit of ~500 GA's
https://knx-user-forum.de/forum/supportforen/cometvisu/1142506-maximale-anzahl-an-gruppenadressen

Suggestions for a solution: move to HTTP POST, it just needs to be changed in CV-Client and to be tested in backend (maybe the buffer-size avoiding DoS in knxd-backend is also too low) which could result in a similar issue in knxd (was eibd; eibd-read/write.cgi.c) which could lead to another issue in knxd..

But basically the "protocol" was designed for more than that, so we should find a fix :)

**Comment by peuter** [comment]
Just changing the read request from GET to POST would the easiest solution, but not the way we should got here for at least two reasons.

- POST requests should be use to send data to a Server not for retrieving data from it
- we support different transport layers in the CometVisu and at least the "Server-sent-events" will not work with POST requests.

Basically we should stick to the protocol here by implementing filters:
https://github.com/CometVisu/CometVisu/wiki/Protokoll#Filter

Needs to be implemented on client and backend side.

**Comment by ChristianMayer** [comment]
I'd suggest to enable the backend to accept GET and POST and return in `l` a version of 0.0.2.
At the same time the client could be enhanced to use GET with a version of 0.0.1 (the current standard) and POST with 0.0.2. All of this should be very simple.

A bit bigger task would be to enhance the backend to support also filters (according to the spec update), use a higher version (0.1.0? Or even 1.0.0?) and implement it in the client accordingly.
The big issue with this step is that the backend has to switch from the current state less implementation to a state full as it must store the filter information somewhere.
This can be implemented in two ways, the likely simple one is still sticking with CGI and use e.g. `/tmp/` or `/var/` to store the currently known filters.
The more advanced one would create a deamon out of it that e.g. talks FastCGI with the web server process and stores the known filters in RAM.

(All of that is on my TODO list - but the priority is/was not very high as the current implementation used to work for everyone till now)

**Comment by Makki1** [comment]
@ChristianMayer : I fully agree; moving to POST should be rather easy, if I remember my quirky source-code on the knxd-backend correctly it's even maybe >90% already implemented. (because we talked about that years ago without having - but already seeing the possible problem..)
Maybe I oversee something but IMHO it doesn't even require to change/distinguish by "protocol version" (?!)
Chris stated it correctly: changing from the stateless CGI to a stateful deamon with error-prone "filter" which requires to be stateful on the server/backend is a much bigger effort and IMHO a rather useless one.
It was initially simply designed the other way around: being very lean on Client *and* Server without having to keep track of a hundred corner-cases that would be introduced by a stateful daemon.
I don't see any benefit in changing that just for the sake of being more complex to handle(?) and much more resource-intense on the stateful server/backend;
Simple use-case: Think about: you might have 100 Clients. Currently each one keeps track where they are, the backend just doesn't need to take care about that. 50 of them maybe go offline occiasionally, no problem currently.
With a stateful filter on the server this will be a problem!
(Also I don't see, how "filter" resolves the underlying problem in any way, if I [the frontend] just wants updates for any&all GA's?! which might be legitim)
When do I "drop" Clients (Filters) on the Server? 1min, 1h, 1d?
Think about it! Dunno, DoS is open for anyone then.. The current design is (after some 8 yrs) still pretty immune to DoS-Attacks. 

Anyway, if - it needs to be implemented in knxd, not in the eibread/write CGI, like it already is with i=xxxxx. But why? It works, the only problem is the size-limitation of HTTP-GET
The protocol is IMHO still very good, lean, efficient, it just has to be understood the right way.

@peuter I dont see why using POST instead of GET should be "bad" - it's an easy solution for an existing problem without any harm. 
Other backends could handle it different if it makes sense but thats a another story then..

**Comment by joltcoke** [comment]
The key question here is: Do we want a common protocol, which is the same for all backends or do we create backend specific protocols?

**Comment by peuter** [comment]
A complete implementation of the CometVisu-Protocol needs to be stateful anyways. I fully understand that implementing a stateful backend is a huge task, which opens the door for many problems that have to be handled. But as it has to be done anyways in the future, and it would be a clean solution for the existing problem I suggested it here.

If we go the "POST" way, we would have a split protocol here. Lets say version 0.0.2 is with POST instead of GET. If I implement Filters for the openHAB backend (the only way here, because POST is not possible with SSE) we have a 0.0.3 which is incompatible to 0.0.2. Not really nice. Also all those cases need to be handled by the CometVisu-Client.

@Makki1 using POST instead of GET is at least bad style in this case. But thats more an academic problem. The main problem is, that it would be a backend-specific solution, which will not apply to the openHAB backend. Despite the fact that I cannot open a SSE connection with a POST request, an implementation of that pattern will never be accepted as PR. All backends in openHAB are treated as RESTful-APIs and in that area you must use GET to retrieve data and POST to create something new.

So long story short: I still think that sticking to the existing Protocol by using Filters is the way we should go here, but as I will not be the one who is implementing it (at least on the backend side), I leave the decision to whoever is going to do the work ;-) 
I can only support by implementing the client side of the filters, if needed.

**Comment by ChristianMayer** [comment]
We need a "new" version number as otherwise the new client could use POST with the old backend which is looking for GETs.

The reason to implement filters is to get the traffic down on huge installations. The aim should be to have the full client request always in ~1400 bytes... (I guess that's currently about 160 different GAs).
Even more packets is no problem in normal environments. But with the good German mobile network with GPRS even in cities like Munich (O2... :() it's a valid aim to reach...

@joltcoke this is mostly relevant for the classic CometVisu protocol and backend. Other backends that are talking the CometVisu protocol are already able to implement everything and it would still work (as the client isn't using those features yet...). 
Backends that are implementing a different protocol aren't affected by this discussion

**Comment by Makki1** [comment]
Hmm.. Maybe I'm missing something (?)
The basic problem to be solved is that a CV frontend-client wants to be subscribed to more than ~500 GAs of 65k possible (which is already handled very efficient as it still only gets changed ones)
@peuter I don't see how filters would solve this problem in any way? If I (CV) want updates on all of them, then I need to get them (regardless of BW/packet-size).
It's just about telling the server which I want where we hit the GET-limit here..
I still don't understand the benefit of "filter", adding a major complexity to both- backend and frontend -
 compared to the current solution.
The "filter" is simply (without additional "filter") what the FE requests to be sent updates on(?) So "filter" is redundant and useless (and the problem isn't anyway the size of the answer to GET/POST)
If it's academic right or wrong: well, honestly I don't care ;)

**Comment by joltcoke** [comment]
The PUT/POST call to create the filter doesn’t have the size restriction. The GET call (using the filters) to retrieve the status will no longer have size issues.

Usually the backend needs to subscribe for events on the bus. Having static filters available would allow for optimizations there as well.

In fact the current situation is also not entirely stateless as we need to make sure to capture event changes inbetween read calls as well. The protocol tries to hide this by using the timestamp.

**Comment by ChristianMayer** [comment]
The Docker container for the 0.12 release is configured to accept much more GAs now.

**Comment by ChristianMayer** [comment]
By the switch to the new FastCGI backend less severe. And the new backend allows to switch to stateful GA subscription, reducing the communication dramatically
